File: Java/RotateArray.java
package Java;
// Rotate Array in Java
// Problem: Rotate an array of n elements to the right by k steps. For example, 
// with n = 7 and k = 3, the array [1,2,3,4,5,6,7] is rotated to [5,6,7,1,2,3,4].
// How many different ways do you know to solve this problem?
/**
 * RotateArray
 */
public class RotateArray {
    public static void main(String[] args) {
        int[] arr = new int[]{1,2,3,4,5,6,7};   
        int n = arr.length;
        int k = 3;
        // IntermediateArray(arr, k, n);
        // BubbleRotate(arr, k, n);
        Reversal(arr, k, n);
    }
    public static void IntermediateArray(int[] arr,int k,int n){
        // Space is O(n) and time is O(n).
        //Ensure that k is within array bounds
        if(k > n) k %= n;
        //Create a duplicate array with same size of original array
        int[] res = new int[n];
        //Copy last k elements to the result array
        for(int i = 0;i < k;i++){
            res[i] = arr[n-k+i];
        }
        //Add remaining elements from the original array
        int j = 0;
        for(int i = k;i < n;i++){
            res[i] = arr[j];
            j++;
        }
    }
    public static void BubbleRotate(int[] arr,int k,int n){
        for(int i = 0;i < k;i++){
            for(int j = n - 1;j > 0;j--){
                // System.out.println();
                // System.out.println(arr[j]);
                // System.out.println(arr[j-1]);
                int temp = arr[j];
                arr[j] = arr[j-1];
                arr[j-1] = temp;
                // System.out.println("MiniSwap");
                // for (int l : arr) {
                //     System.out.print(l+" ");
                // }
            }
            // System.out.println(" ");
            // System.out.println("MajorSwap");
        }
        for (int i : arr) {
            System.out.print(i+" ");
        }
    }
    public static void Reversal(int[] arr, int k,int n){
        int a = arr.length - k;
        helper(arr, 0, a-1);
        helper(arr, a, arr.length-1);
        helper(arr, 0, arr.length-1);
        for (int i : arr) {
            System.out.print(i+" ");
        }
    }
    private static int[] helper(int[] arr,int left,int right){
            while(left < right){
                int temp = arr[left];
                arr[left] = arr[right];
                arr[right] = temp;
                left++;
                right--;
            }
            return arr;
    } 
}

File: Python/Rotate_Array.py
# // Rotate Array in Java
# // Problem: Rotate an array of n elements to the right by k steps. For example, 
# // with n = 7 and k = 3, the array [1,2,3,4,5,6,7] is rotated to [5,6,7,1,2,3,4].
# // How many different ways do you know to solve this problem?
# /**
#  * RotateArray
#  */
arr = [1,2,3,4,5]
n = len(arr)
k = 2
def IntermediateArray(arr,k,n):
    #Ensure that k is within array bounds
    if k > n:
        k %= n
    #Create a duplicate array wuth same size of original array
    res = []
    #Copy last k elements to the result array
    for i in range(k):
        res.append(arr[n - k + i])
    #Add remaining elements from the original array
    j = 0
    for i in range(k,n):
        res.append(arr[j])
        j+=1
    print(res)
# print(IntermediateArray(arr,k,n))
def Rotate_Array(arr,k,n):
    temp = 0
    for i in range(0,k):
        for j in range(n-1,0,-1):
            # arr[j],arr[j-1] = arr[j-1],arr[j]
            temp =arr[j]
            arr[j] =arr[j-1]
            arr[j-1] = temp
    print(arr)
# Rotate_Array(arr=[1,2,3,4,5,6,7],n=len(arr),k=3)
#This same shit code gives correct output when used somewhere else;((
def Reversal(arr,k,n):
    x = n - k
    def helper(arr,left,right):
        while(left < right):
            arr[left],arr[right] = arr[right],arr[left]
            left+=1
            right-=1
    helper(arr,0,x-1)
    helper(arr,x,n-1)
    helper(arr,0,n-1)
    return arr
print(Reversal(arr=[1,2,3,4,5,6,7],k = 3,n = len(arr)))
#This same shit code gives correct output when used somewhere else;((
